File: DSA/Python/src/dsa/lib/math/ds/paren.py
from dsa.container.lists.queues import Stack


class Parentheses:
    # LeetCode20
    def isValid(self, s: str) -> bool:
        return self.isValid2(s)

    # Stack and Hash Map
    @staticmethod
    def isValid1(s: str) -> bool:
        paren_map = {')': '(', '}': '{', ']': '['}
        stack = Stack()
        for c in s:
            # Current character is an opening bracket.
            if c not in paren_map:
                stack.push_back1(c)
            # Current character is a closing bracket
            elif not stack or stack.pop() != paren_map.get(c):
                return False
        # If the final stack is empty, the string is valid
        return not stack

    # Stack
    @staticmethod
    def isValid2(s: str) -> bool:
        stack: Stack = Stack()
        for c in s:
            # Opening bracket
            if c == '(':
                stack.push_back1(')')
            elif c == '[':
                stack.push_back1(']')
            elif c == '{':
                stack.push_back1('}')
            # Closing bracket
            elif stack.empty() or stack.pop() != c:
                return False
        return stack.empty()

File: DSA/Cpp/src/dsa/lib/math/ds/paren.cpp
#include <stack>
#include <map>
#include "paren.h"

namespace dsa::lib::math::ds {

bool Parentheses::isValid(const std::string &s) {
    return isValid1(s);
}
//Stack and Hash Map
bool Parentheses::isValid1(const std::string &s) {
    std::stack<char> stack;
    std::map<char, char> parenMap = {{']', '['}, {')', '('}, {'}', '{'}};
    for (const char &c : s) {
//        Opening bracket. Did not find the key
        if (parenMap.find(c) == parenMap.end()) stack.push(c);
//        Closing bracket
        else if (stack.empty() || stack.top() != parenMap[c]) return false;
        else stack.pop();
    }
    return stack.empty();
}
//Stack
bool Parentheses::isValid2(const std::string &s) {
    std::stack<char> stack;
    for (const auto &c: s) {
//        Opening bracket
        if (c == '(') stack.push(')');
        else if (c == '[') stack.push(']');
        else if (c == '{') stack.push('}');
//        Closing bracket
        else if (stack.empty() || stack.top() != c) return false;
        else stack.pop();
    }
    return stack.empty();
}
bool Parentheses::isValid21(const std::string &s) {
    std::stack<char> stack;
    for (const auto &c : s) {
        switch (c) {
            case '(':stack.push(')');
                break;
            case '[': stack.push(']');
                break;
            case '{': stack.push('}');
                break;
            default:
                if (stack.empty() || stack.top() != c) return false;
                else stack.pop();
        }
    }
    return stack.empty();
}

}
